From 796085825c4c4df77ca730be25b43ff2052add9d Mon Sep 17 00:00:00 2001
From: S-trace <S-trace@list.ru>
Date: Sun, 3 Apr 2016 17:35:17 +0300
Subject: [PATCH] [TORCH] Use sysfs interface for SystemUI FlashlightController

This should improve stability, performance and battery consumption of Flashlight QuickSettings title

Change-Id: I65cd7a07f559d3e7f4cf505c164e2b50f50283cb
---
 .../statusbar/policy/FlashlightController.java     | 44 ++++++++++++++--------
 1 file changed, 29 insertions(+), 15 deletions(-)

diff --git a/packages/SystemUI/src/com/android/systemui/statusbar/policy/FlashlightController.java b/packages/SystemUI/src/com/android/systemui/statusbar/policy/FlashlightController.java
index d307a4ceb78a..4b11fc79676f 100644
--- a/packages/SystemUI/src/com/android/systemui/statusbar/policy/FlashlightController.java
+++ b/packages/SystemUI/src/com/android/systemui/statusbar/policy/FlashlightController.java
@@ -35,6 +35,9 @@ import android.util.Log;
 
 import com.android.systemui.R;
 
+import java.io.FileWriter;
+import java.io.IOException;
+
 import java.io.FileDescriptor;
 import java.io.PrintWriter;
 import java.lang.ref.WeakReference;
@@ -52,10 +55,16 @@ public class FlashlightController {
     private static final int DISPATCH_CHANGED = 1;
     private static final int DISPATCH_AVAILABILITY_CHANGED = 2;
 
+    public static final int OFF  = 0;
+    public static final int ON   = 127;
+    public static final int HIGH = 255;
+
+    private FileWriter mFlashDeviceWriter = null;
+
     private static final String ACTION_TURN_FLASHLIGHT_OFF =
             "com.android.systemui.action.TURN_FLASHLIGHT_OFF";
 
-    private final CameraManager mCameraManager;
+    private final CameraManager mCameraManager = null;
     private final Context mContext;
     /** Call {@link #ensureHandler()} before using */
     private Handler mHandler;
@@ -66,8 +75,8 @@ public class FlashlightController {
     /** Lock on {@code this} when accessing */
     private boolean mFlashlightEnabled;
 
-    private String mCameraId;
-    private boolean mTorchAvailable;
+    private final String mCameraId = null;
+    private boolean mTorchAvailable = true;
 
     private Notification mNotification = null;
     private boolean mReceiverRegistered;
@@ -89,22 +98,20 @@ public class FlashlightController {
 
     public FlashlightController(Context context) {
         mContext = context;
-        mCameraManager = (CameraManager) mContext.getSystemService(Context.CAMERA_SERVICE);
+        Log.d(TAG, "Initializing");
 
-        tryInitCamera();
+        tryInitWriter();
     }
 
-    private void tryInitCamera() {
+    private void tryInitWriter() {
+        FileWriter FlashDeviceWriter = null;
         try {
-            mCameraId = getCameraId();
+            FlashDeviceWriter = new FileWriter("/sys/class/leds/torch-light/brightness");
         } catch (Throwable e) {
             Log.e(TAG, "Couldn't initialize.", e);
             return;
-        }
-
-        if (mCameraId != null) {
-            ensureHandler();
-            mCameraManager.registerTorchCallback(mTorchCallback, mHandler);
+        } finally {
+            mFlashDeviceWriter = FlashDeviceWriter ;
         }
     }
 
@@ -116,8 +123,15 @@ public class FlashlightController {
                 mFlashlightEnabled = enabled;
 
                 try {
-                    mCameraManager.setTorchMode(mCameraId, enabled);
-                } catch (CameraAccessException e) {
+                    if (mFlashDeviceWriter != null) {
+                        int level = enabled ? ON : OFF;
+                        Log.d(TAG, "Setting torch level to " + String.valueOf(level));
+                        mFlashDeviceWriter.write(String.valueOf(level));
+                        mFlashDeviceWriter.flush();
+                    } else {
+                        Log.e(TAG, "mFlashDeviceWriter == null!!");
+                    }
+                } catch (Exception e) {
                     Log.e(TAG, "Couldn't set torch mode", e);
                     mFlashlightEnabled = false;
                     pendingError = true;
@@ -195,7 +209,7 @@ public class FlashlightController {
     public void addListener(FlashlightListener l) {
         synchronized (mListeners) {
             if (mCameraId == null) {
-                tryInitCamera();
+                tryInitWriter();
             }
             cleanUpListenersLocked(l);
             mListeners.add(new WeakReference<>(l));
-- 
2.11.0

